_posts/2024-10-30-lower-your-cost-on-opensearch-using-binary-vectors.md

```diff
@@ -0,0 +1,155 @@
+
+<img src="/assets/media/blog-images/2024-10-30-lower-your-cost-on-opensearch-using-binary-vectors/pic1.png" alt="fp32-binary-comparison" class="img-centered"/>
+
+The binary format makes binary vectors more compact and faster to process. LLMs are increasingly generating binary embeddings for their efficiency on large datasets, offering significant reductions in storage, memory, and latency.
+
+## Using binary vectors in OpenSearch
+
+Let’s explore how you can use binary vectors in your OpenSearch solution.
+
+### Data preparation
+
+To get started, you’ll need binary vector data. Luckily, many models now generate embeddings in binary format. For example, the Cohere Embed v3 model generates binary vectors.
+
+Binary vectors are arrays of zeros and ones, such as `[0, 1, 1, 0, 1, 1, 0, 0, 1, 0, 0, 0, 1, 1, 0, 0]`. However, OpenSearch requires binary vectors packed into an `int8` byte format. For example, the preceding bit array would be converted to `[108, -116]`, as shown in the following image.
+
+<img src="/assets/media/blog-images/2024-10-30-lower-your-cost-on-opensearch-using-binary-vectors/pic2.png" alt="binary-vector-packing" class="img-centered"/>
+
+Many embedding models already generate binary vectors in `int8` byte format, so extra packing is usually unnecessary. However, if your data is stored in a bit array, you can easily convert it into a byte array using the `numpy` library:
+
+```python
+import numpy as np
+bit_array = [0, 1, 1, 0, 1, 1, 0, 0, 1, 0, 0, 0, 1, 1, 0, 0] 
+bit_array_np = np.array(bit_array, dtype=np.uint8) 
+byte_array = np.packbits(bit_array_np).astype(np.int8).tolist()
+```
+
+### Ingestion and search
+
+Once you have byte arrays storing your data, you'll need to ingest those into OpenSearch. 
+
+First, set the data type to `binary` in your index mapping and ensure the vector dimensions are a multiple of 8 (if they are not, pad the vectors with zeros):
+
+```json
+PUT /test-binary-hnsw
+{
+  "settings": {
+    "index": {
+      "knn": true
+    }
+  },
+  "mappings": {
+    "properties": {
+      "my_vector": {
+        "type": "knn_vector",
+        "dimension": 16,
+        "data_type": "binary",
+        "space_type": "hamming",
+        "method": {
+          "name": "hnsw",
+          "engine": "faiss"
+        }
+      }
+    }
+  }
+}
+```
+
+Binary vectors in OpenSearch use Hamming distance for indexing and search.
+
+Next, pack the binary vector into a byte format for both indexing and searching. Other than that, using binary vectors is similar to working with FP32 vectors. In the following example, you'll index two documents with vector values of `[0, 0, 0, 0, 0, 1, 1, 1, 0, 0, 0, 0, 1, 0, 0, 0]` and `[0, 0, 0, 0, 1, 0, 1, 0, 0, 0, 0, 0, 1, 0, 1, 1]`:
```

Comment:
```suggestion
Next, pack the binary vector into a byte format for both indexing and searching. Otherwise, using binary vectors is similar to working with FP32 vectors. In the following example, you'll index two documents with vector values of `[0, 0, 0, 0, 0, 1, 1, 1, 0, 0, 0, 0, 1, 0, 0, 0]` and `[0, 0, 0, 0, 1, 0, 1, 0, 0, 0, 0, 0, 1, 0, 1, 1]`:
```